datalad/customremotes/ria_remote.py

```diff
@@ -0,0 +1,933 @@
+        if not archive_path.exists():
+            # no archive, not file
+            return False
+        loc = str(file_path)
+        from datalad.cmd import Runner
+        runner = Runner()
+        # query 7z for the specific object location, keeps the output
+        # lean, even for big archives
+        out, err = runner(
+            ['7z', 'l', str(archive_path),
+             loc],
+            log_stdout=True,
+        )
+        return loc in out
+
+    def read_file(self, file_path):
+
+        with open(str(file_path), 'r') as f:
+            content = f.read()
+        return content
+
+    def write_file(self, file_path, content, mode='w'):
+
+        with open(str(file_path), mode) as f:
+            f.write(content)
+
+
+class SSHRemoteIO(IOBase):
+    """IO operation if the object tree is SSH-accessible
+
+    It doesn't even think about a windows server.
+    """
+
+    # output markers to detect possible command failure as well as end of output from a particular command:
+    REMOTE_CMD_FAIL = "ria-remote: end - fail"
+    REMOTE_CMD_OK = "ria-remote: end - ok"
+
+    def __init__(self, host):
+        """
+        Parameters
+        ----------
+        host : str
+          SSH-accessible host(name) to perform remote IO operations
+          on.
+        """
+
+        from datalad.support.sshconnector import SSHManager
+        # connection manager -- we don't have to keep it around, I think
+        self.sshmanager = SSHManager()
+        # the connection to the remote
+        # we don't open it yet, not yet clear if needed
+        self.ssh = self.sshmanager.get_connection(
+            host,
+            use_remote_annex_bundle=False,
+        )
+        self.ssh.open()
+        # open a remote shell
+        cmd = ['ssh'] + self.ssh._ssh_args + [self.ssh.sshri.as_str()]
+        self.shell = subprocess.Popen(cmd, stderr=subprocess.DEVNULL, stdout=subprocess.PIPE, stdin=subprocess.PIPE)
+        # swallow login message(s):
+        self.shell.stdin.write(b"echo RIA-REMOTE-LOGIN-END\n")
+        self.shell.stdin.flush()
+        while True:
+            line = self.shell.stdout.readline()
+            if line == b"RIA-REMOTE-LOGIN-END\n":
+                break
+        # TODO: Same for stderr?
+
+    def close(self):
+        # try exiting shell clean first
+        self.shell.stdin.write(b"exit\n")
+        self.shell.stdin.flush()
+        exitcode = self.shell.wait(timeout=0.5)
+        # be more brutal if it doesn't work
+        if exitcode is None:  # timed out
+            # TODO: Theoretically terminate() can raise if not successful. How to deal with that?
+            self.shell.terminate()
+        self.sshmanager.close()
+
+    def _append_end_markers(self, cmd):
+        """Append end markers to remote command"""
+
+        return cmd + ' && echo "{}" || echo "{}"\n'.format(self.REMOTE_CMD_OK, self.REMOTE_CMD_FAIL)
+
+    def _get_download_size_from_key(self, key):
```

Comment:
How is this different from `AnnexRepo.get_size_from_key()`? It only says "doesn't fit", but now why or how. If annexrepo import is too expensive, it might be worth filing an issue that there is code duplication happening and that we should RF accordingly. `AnnexRepo.get_size_from_key()` is a static method, but could easily move to `datalad.utils` as a standalone function.

Reply:
This is outdated. I already had a PR fixing `AnnexRepo`'s method in preparation of the RF.

Reply:
Does that mean the function is outdated and can be removed, or the comment is outdated and should be removed?